File: ios-app/Sources/NativeCheckpointUI/ExactNavigationCheckpoint.swift
import NativeParityDomain
import SwiftUI

public enum ExactNavigationAccessibility {
    public static let root = "exact-navigation-root"
    public static let state = "exact-navigation-state"
    public static let approve = "exact-navigation-approve"
}

public enum DispatchStatus: String, Equatable {
    case accepted = "ACCEPTED"
    case duplicate = "CHECKPOINT_DUPLICATE_DISPATCH"
}

public struct DispatchReceipt: Equatable {
    public let status: DispatchStatus
    public let operationID: String
}

public protocol NavigationEffectPort: AnyObject {
    func dispatch(_ command: NavigationCommand) -> DispatchReceipt
}

public final class RecordingNavigationEffectPort: NavigationEffectPort {
    private var consumedOperationIDs: Set<String> = []
    public private(set) var commands: [NavigationCommand] = []

    public init() {}

    public func dispatch(_ command: NavigationCommand) -> DispatchReceipt {
        guard consumedOperationIDs.insert(command.operationID).inserted else {
            return DispatchReceipt(status: .duplicate, operationID: command.operationID)
        }
        commands.append(command)
        return DispatchReceipt(status: .accepted, operationID: command.operationID)
    }
}

public final class ExactNavigationCheckpointController {
    private let gate: NavigationGate
    private let effectPort: NavigationEffectPort

    public var state: GateState { gate.state }

    public init(gate: NavigationGate, effectPort: NavigationEffectPort) {
        self.gate = gate
        self.effectPort = effectPort
    }

    @discardableResult
    public func propose(operationID: String, targetPageID: String, sourceGeneration: Int64) -> Bool {
        gate.propose(
            NavigationProposal(
                operationID: operationID,
                targetPageID: targetPageID,
                sourceGeneration: sourceGeneration
            )
        )
    }

    @discardableResult
    public func approve(operationID: String) -> Bool {
        guard let command = gate.approve(operationID: operationID) else { return false }
        return effectPort.dispatch(command).status == .accepted
    }

    public func pageIdentityChanged(pageID: String, generation: Int64) {
        gate.pageIdentityChanged(pageID: pageID, generation: generation)
    }

    @discardableResult
    public func callbackAndObserve(
        operationID: String,
        accepted: Bool,
        landedPageID: String
    ) -> GateState {
        guard gate.platformCallback(operationID: operationID, accepted: accepted) else {
            return gate.state
        }
        return gate.verify(operationID: operationID, landedPageID: landedPageID)
    }
}

public struct ExactNavigationCheckpointView: View {
    private let state: GateState
    private let onApprove: () -> Void

    public init(state: GateState, onApprove: @escaping () -> Void) {
        self.state = state
        self.onApprove = onApprove
    }

    public var body: some View {
        VStack {
            Text("State: \(state.rawValue)")
                .accessibilityIdentifier(ExactNavigationAccessibility.state)
            Button("Approve", action: onApprove)
                .accessibilityIdentifier(ExactNavigationAccessibility.approve)
        }
        .accessibilityIdentifier(ExactNavigationAccessibility.root)
    }
}

File: android-app/src/main/kotlin/dev/ed3c/nativeparity/checkpoint/ExactNavigationCheckpoint.kt
package dev.ed3c.nativeparity.checkpoint

import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.text.BasicText
import androidx.compose.runtime.Composable
import androidx.compose.ui.Modifier
import androidx.compose.ui.semantics.contentDescription
import androidx.compose.ui.semantics.semantics
import dev.ed3c.nativeparity.GateState
import dev.ed3c.nativeparity.NavigationCommand
import dev.ed3c.nativeparity.NavigationGate
import dev.ed3c.nativeparity.NavigationProposal

object ExactNavigationAccessibility {
    const val ROOT = "exact-navigation-root"
    const val STATE = "exact-navigation-state"
    const val APPROVE = "exact-navigation-approve"
}

enum class DispatchStatus {
    ACCEPTED,
    CHECKPOINT_DUPLICATE_DISPATCH,
}

data class DispatchReceipt(
    val status: DispatchStatus,
    val operationId: String,
)

fun interface NavigationEffectPort {
    fun dispatch(command: NavigationCommand): DispatchReceipt
}

class RecordingNavigationEffectPort : NavigationEffectPort {
    private val consumedOperationIds = mutableSetOf<String>()
    val commands = mutableListOf<NavigationCommand>()

    override fun dispatch(command: NavigationCommand): DispatchReceipt {
        if (!consumedOperationIds.add(command.operationId)) {
            return DispatchReceipt(DispatchStatus.CHECKPOINT_DUPLICATE_DISPATCH, command.operationId)
        }
        commands += command
        return DispatchReceipt(DispatchStatus.ACCEPTED, command.operationId)
    }
}

class ExactNavigationCheckpointController(
    private val gate: NavigationGate,
    private val effectPort: NavigationEffectPort,
) {
    val state: GateState
        get() = gate.state

    fun propose(operationId: String, targetPageId: String, sourceGeneration: Long): Boolean =
        gate.propose(NavigationProposal(operationId, targetPageId, sourceGeneration))

    fun approve(operationId: String): Boolean {
        val command = gate.approve(operationId) ?: return false
        return effectPort.dispatch(command).status == DispatchStatus.ACCEPTED
    }

    fun pageIdentityChanged(pageId: String, generation: Long) {
        gate.pageIdentityChanged(pageId, generation)
    }

    fun callbackAndObserve(operationId: String, accepted: Boolean, landedPageId: String): GateState {
        if (!gate.platformCallback(operationId, accepted)) return gate.state
        return gate.verify(operationId, landedPageId)
    }
}

@Composable
fun ExactNavigationCheckpointScreen(
    state: GateState,
    onApprove: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Column(modifier.semantics { contentDescription = ExactNavigationAccessibility.ROOT }) {
        BasicText(
            text = "State: ${state.name}",
            modifier = Modifier.semantics {
                contentDescription = ExactNavigationAccessibility.STATE
            },
        )
        BasicText(
            text = "Approve",
            modifier = Modifier
                .semantics { contentDescription = ExactNavigationAccessibility.APPROVE }
                .clickable(onClick = onApprove),
        )
    }
}
